doc/plutus-core-spec/untyped-grammar.tex

```diff
@@ -64,15 +65,17 @@ \subsection{Grammar}
 \begin{minipage}{\linewidth}
     \centering
     \[\begin{array}{lrclr}
-    \textrm{Term}       & L,M,N  & ::= & x                      & \textrm{variable}\\
-                        &        &     & \con{\tn}{c}           & \textrm{constant}\\
-                        &        &     & \builtin{b}            & \textrm{builtin}\\
-                        &        &     & \lamU{x}{M}            & \textrm{$\lambda$ abstraction}\\
-                        &        &     & \appU{M}{N}            & \textrm{function application}\\
-                        &        &     & \delay{M}              & \textrm{delay execution of a term}\\
-                        &        &     & \force{M}              & \textrm{force execution of a term}\\
-                        &        &     & \errorU                & \textrm{error}\\
-        \textrm{Program}& P      & ::= & \version{v}{M}         & \textrm{versioned program}
+    \textrm{Term}       & L,M,N  & ::= & x                               & \textrm{variable}\\
+                        &        &     & \con{\tn}{c}                    & \textrm{constant}\\
+                        &        &     & \builtin{b}                     & \textrm{builtin}\\
+                        &        &     & \lamU{x}{M}                     & \textrm{$\lambda$ abstraction}\\
+                        &        &     & \appU{M}{N}                     & \textrm{function application}\\
+                        &        &     & \delay{M}                       & \textrm{delay execution of a term}\\
+                        &        &     & \force{M}                       & \textrm{force execution of a term}\\
+                        &        &     & \constr{i}{M_0 \ldots M_{m-1}}  & \textrm{constructor with tag $i$ and $m$ arguments}\\
```

Comment:
I note that later on you usually use M_0 ... M_n, forgetting about the -1.  Maybe you can just do that everywhere since the exact number of arguments doesn't matter.  

Oh, but wait, it's perfectly OK to have no arguments at all and then you probably need m=0 so that m-1 gives you an empty list.  Maybe `constr` needs m-1 elsewhere because of that.  I think the document implicitly assumes list indices are positive, but if we allow m<0 then M_1 ... M_m would be empty.

Reply:
Yeah, the tags start at 0, so it would be annoying not to start at 0 with the numbering here.

Elsewhere I've use `M_0 ... M_n` to just be "some things where we don't really care how many there are". The only reason I used `M_{n-1}` here is so I could say "a constructor with m arguments". I could just not say that and go the other way, relying on the reader to count.

I think it's a bit ambiguous what is going on in the empty case, but I don't think it's confusing? I think people often write `x_0 ... x_n` and implicitly intend to include the empty case? IDK.

Reply:
I'll think about which lists can be empty and maybe add a sentence for clarification somewhere.